src/Model/Aggregate.php

```diff
@@ -0,0 +1,299 @@
+            $this->aggregateExpressions[$name] = $expr;
+
+            $seed = is_array($expr) ? $expr : [$expr];
+
+            $args = [];
+            // if field originally defined in the parent model, then it can be used as part of expression
+            if ($this->baseModel->hasField($name)) {
+                $args = [$this->baseModel->getField($name)];
+            }
+
+            $seed['expr'] = $this->baseModel->expr($seed[0] ?? $seed['expr'], $args);
+
+            // now add the expressions here
+            $this->addExpression($name, $seed);
+        }
+
+        return $this;
+    }
+
+    /**
+     * Return reference field.
+     *
+     * @param string $link
+     */
+    public function getRef($link): Reference
+    {
+        return $this->baseModel->getRef($link);
+    }
+
+    /**
+     * Method to enable commutative usage of methods enabling both of below
+     * Resulting in Aggregate on $model.
+     *
+     * $model->groupBy(['abc'])->withAggregateField('xyz');
+     *
+     * and
+     *
+     * $model->withAggregateField('xyz')->groupBy(['abc']);
+     */
+    public function withAggregateField(string $name, $seed = []): Model
+    {
+        static::addField($name, $seed);
+
+        return $this;
+    }
+
+    /**
+     * Adds new field into model.
+     *
+     * @param array|object $seed
+     */
+    public function addField(string $name, $seed = []): Field
+    {
+        $seed = is_array($seed) ? $seed : [$seed];
+
+        if (isset($seed[0]) && $seed[0] instanceof FieldSqlExpression) {
+            return parent::addField($name, $seed[0]);
+        }
+
+        if ($seed['never_persist'] ?? false) {
+            return parent::addField($name, $seed);
+        }
+
+        if ($this->baseModel->hasField($name)) {
+            $field = clone $this->baseModel->getField($name);
+            $field->unsetOwner(); // will be new owner
+        } else {
+            $field = null;
+        }
+
+        return $field
+            ? parent::addField($name, $field)->setDefaults($seed)
+            : parent::addField($name, $seed);
+    }
+
+    public function setLimit(int $count = null, int $offset = 0)
+    {
+        $this->baseModel->setLimit($count, $offset);
+
+        return $this;
+    }
+
+    /**
+     * @param string $mode
+     * @param array  $args
+     *
+     * @return Query
+     */
+    public function action($mode, $args = [])
+    {
+        switch ($mode) {
+            case 'select':
+                $fields = $this->only_fields ?: array_keys($this->getFields());
+
+                // select but no need your fields
+                $query = $this->baseModel->action($mode, [false]);
+
+                $this->initQueryFields($query, array_unique($fields + $this->groupByFields));
+                $this->initQueryOrder($query);
+                $this->initQueryGrouping($query);
+                $this->initQueryConditions($query);
+
+                $this->hook(self::HOOK_INIT_SELECT_QUERY, [$query]);
```

Comment:
why to not use `Persistence\Sql::HOOK_INIT_SELECT_QUERY` (just neutral question, no sudgestion for now)

Reply:
TODO after https://github.com/atk4/data/pull/946